samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java

```diff
@@ -75,16 +75,16 @@
 
   private static final String JAVA_OPTS_CONTAINER_NAME = "samza.container.name";
   private static final String JOB_COORDINATOR_TAG = "samza-job-coordinator";
-  private static final String SOURCE = "log4j-log";
+  protected static final String SOURCE = "log4j-log";
 
   // Hidden config for now. Will move to appropriate Config class when ready to.
   private static final String CREATE_STREAM_ENABLED = "task.log4j.create.stream.enabled";
 
   private static final long DEFAULT_QUEUE_TIMEOUT_S = 2; // Abitrary choice
   private final BlockingQueue<byte[]> logQueue = new LinkedBlockingQueue<>(DEFAULT_QUEUE_SIZE);
 
-  private SystemStream systemStream = null;
-  private SystemProducer systemProducer = null;
+  protected SystemStream systemStream = null;
+  protected SystemProducer systemProducer = null;
```

Comment:
protected fields should be grouped with other protected fields.

Reply:
can you help me understand how subclasses are going to use these three fields?

Reply:
a subclass can opt to not have partition key and record key when sending out the message to the underlying system producer based on how the underlying logging system distributes the messages across partitions. 
the #1470 pr exposed the method sending to the systemproducer.

for example: if the partition key and record key are null in the OutgoingMessageEnvelope, kafka will use round robin instead of key hash to find the partition. this will prevent partitions with TB of data while rest of partitions within same topic are empty due to the hash.